java-checks/src/main/java/org/sonar/java/checks/spring/AvoidQualifierOnBeanMethodsCheck.java

```diff
@@ -0,0 +1,117 @@
+  private static final String BEAN_ANNOTATION = "org.springframework.context.annotation.Bean";
+  private static final String QUALIFIER_ANNOTATION = "org.springframework.beans.factory.annotation.Qualifier";
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.METHOD);
+  }
+
+  /**
+   * This rule reports an issue when @Bean methods are annotated with @Qualifier.
+   */
+  @Override
+  public void visitNode(Tree tree) {
+    var methodTree = (MethodTree) tree;
+
+    var beanAnnotation = getAnnotation(methodTree, BEAN_ANNOTATION);
+    var qualifierAnnotation = getAnnotation(methodTree, QUALIFIER_ANNOTATION);
+
+    if(beanAnnotation != null && qualifierAnnotation != null) {
+      QuickFixHelper.newIssue(context)
+       .forRule(this)
+        .onTree(qualifierAnnotation)
+        .withMessage("Remove this redundant \"@Qualifier\" annotation and rely on the @Bean method.")
+        .withQuickFixes(() -> getQuickFix(methodTree, qualifierAnnotation))
+       .report();
+    }
+  }
+
+  private static AnnotationTree getAnnotation(MethodTree methodTree, String annotation) {
+    return methodTree.modifiers()
+      .annotations()
+      .stream()
+      .filter(annotationTree -> annotationTree.symbolType().is(annotation))
+      .findFirst()
+      .orElse(null);
+  }
+
+  private static List<JavaQuickFix> getQuickFix(MethodTree methodTree, AnnotationTree qualifierAnnotation) {
+    List<JavaQuickFix> quickFixes = new LinkedList<>();
+
+    // quick fix only for @Qualifier annotations without arguments or with argument that matches the method name
+    if(isFixable(methodTree, qualifierAnnotation)) {
+      var quickFix = JavaQuickFix.newQuickFix("Remove \"@Qualifier\"")
+        .addTextEdit(JavaTextEdit.removeTree(qualifierAnnotation))
+        .build();
+      quickFixes.add(quickFix);
+    }
+
+    return quickFixes;
+  }
+
+  private static boolean isFixable(MethodTree methodTree, AnnotationTree qualifierAnnotation) {
+    var arguments = qualifierAnnotation.arguments();
+
+    // @Qualifier annotation without argument can be always removed
+    if(arguments.isEmpty()) {
+      return true;
+    }
+
+    // @Qualifier that matches the method name is redundant and can be removed
+    var methodName = methodTree.simpleName().name();
+    return getQualifierAnnotationValue(arguments).equals(methodName);
+  }
+
+  private static String getQualifierAnnotationValue(Arguments arguments) {
+    var argument = arguments.get(0);
+    var qualifierAnnotationValue = "";
+
+    if(argument.is(Tree.Kind.ASSIGNMENT)) {
+      qualifierAnnotationValue = ((LiteralTreeImpl) ((AssignmentExpressionTreeImpl) argument).expression()).value().replace("\"", "");
+    }
+    else if(argument.is(Tree.Kind.STRING_LITERAL)) {
+      qualifierAnnotationValue = ((LiteralTreeImpl) argument).token().text().replace("\"", "");
+    }
+
+    return qualifierAnnotationValue;
```

Comment:
I understand checking the types twice just to be sure but in practice I am not sure if we are likely to see other types of arguments. Have you encountered any other while testing?